### ISSUE TITLE
RuleTester - Add "idempotent" flag for code fixes

### ISSUE THREAD

**Issue by Skalman** [eslint/eslint#13631]
**The version of ESLint you are using.** 7.7.0

**The problem you want to solve.**
I want to make sure that my rule's fixes don't require further fixes, i.e. that once my rule has fixed the code, it won't trigger another error.

Hypothetically, suppose that `semi-spacing` was incorrectly implemented such that:

```js
// Before
var a = "b";var c = "d";

// After first --fix
var a = "b"; var c = "d";

// After second --fix
var a = "b" ; var c = "d";
```

This flag would ensure that the first and second `--fix` always yield the same result, and that the second time would not report any errors. I.e., the fix is idempotent.

Perhaps the implementers hadn't added extensive tests. This `idempotent` flag would automatically make the test suite more comprehensive.

**Your take on the correct solution to problem.**

I'd like to have the `idempotent` flag be on by default, but that would be a breaking change. A disadvantage of having it on by default would be that tests with fixes run slower (is this a concern?).

Anyway, assuming it's off by default for now, it'd look like this:

```js
ruleTester.run("semi-spacing", semiSpacing, {
  valid: [
    // Valid rules aren't affected
  ],

  invalid: [
    {
      // Only rules with an output can be idempotent
      code: 'var a = "b";var c = "d";',
      output: 'var a = "b" ; var c = "d";',
      errors: 1,
      idempotent: true,
    },
  ],
});
```

Names for this flag:

- `idempotent` - would only check that there are no further fixes available
- `outputIsValid` - would check that there are no more errors in the output

An alternative, inferior solution (perhaps user-space) would be to collect all invalid test cases, and add their outputs as "valid" test cases.

**Are you willing to submit a pull request to implement this change?**
Maybe

**Comment by anikethsaha** [comment]
Thanks for the issue. 

I am not able to reproduce the case you mentioned. 
If thr code is correct that is after the first fix, it would be a valid code so the fixer won't run over that. 

```
var a = "b"; var c = "d";
```
This is a valid code and it won't be fix again as this is not an error 
A particular rule will not report a code after that rule fixed that, if that is the case ( which I haven't experience as of now in core rules) there might be some error with the rule itself. 
It can be possible with two different rule reporting each others or either one 's fixed code. 
Also about the flag, in rule tester, this flag won't be usable as you can only have one input and fixed output. So if your input is invalid and fixable, there would be only one fixable code for that. 

Let me know if I am missing something

**Comment by Skalman** [comment]
@anikethsaha, you're misinterpreting the whole issue. This is **not** a bug report, it's a feature request. AFAIK, there isn't any problem with `semi-spacing`, it's only a hypothetical example. :smile:

I've tried to clarify the description. I'm asking for a new flag for `RuleTester`.

**Comment by mdjermanovic** [comment]
Makes sense to me.

We are already running the rule on the output, to check if the autofix has produced code with a parsing error:

https://github.com/eslint/eslint/blob/e60ec07fad0c1d4c966f28d214c5379da753ff4e/lib/rule-tester/rule-tester.js#L563-L571

**Comment by mdjermanovic** [comment]
> This flag would ensure that the first and second `--fix` always yield the same result, and that the second time would not report any errors. I.e., the fix is idempotent.

Producing the same code after two iterations doesn't have to be equivalent to "the second time would not report any errors".
The rule could fix the original code, and then report a non-auto-fixable error in the output. In that case, the output after the second iteration would be still the same.

I think it would be more clear if the option would just check if there are no linting errors in the output.

**Comment by Skalman** [comment]
@mdjermanovic, the parsing error check has saved me a few times already!

I was thinking that an idempotency check could be enabled by default (in the next major).
Checking that there aren't any linting errors in the output can't be enabled by default, but might be more useful in the normal case.

For simplicity, it might be a bad idea to have both options.

Which syntax would you prefer? I can think of two different ones that both have their advantages and disadvantages.

```js
// Specified per test case
ruleTester.run("semi-spacing", semiSpacing, {
  valid: [ /* ... */],
  invalid: [{
    code: 'var a = "b";var c = "d";',
    output: 'var a = "b" ; var c = "d";',
    errors: 1,
    fixesAllErrors: true,
  }],
});

// Specified as its own group
ruleTester.run("semi-spacing", semiSpacing, {
  valid: [/* ... */],
  invalid: [/* ... */],
  invalidWithCompleteFixes: [{
    code: 'var a = "b";var c = "d";',
    output: 'var a = "b" ; var c = "d";',
    errors: 1,
  }],
});
```

I'd appreciate suggestions for better option names :smile:

**Comment by mdjermanovic** [comment]
I'd say the first syntax, and that the option's name should contain the word "output" to make it clear what's exactly being checked if the option was set (the fact that RuleTester doesn't work like `--fix` but runs only one auto-fixing iteration is already confusing).

Maybe something like `hasLintingErrorsInOutput` (or something shorter).

* `false` fails the test if there are linting errors in the output.
* `true` fails the test if there are no linting errors in the output.
* `undefined` (not set) ignores this check.

Fatal (parsing) error always fails, regardless of this option.

This is just an initial idea, there might be a better design.

**Comment by Skalman** [comment]
I mostly agree, but I'd like to reverse the option. What about `outputIsValid` or `isOutputValid`?

**Comment by Skalman** [comment]
@mdjermanovic, I don't understand how the ESLint project works.

This issue is currently tagged as "evaluating", described as "The team will evaluate this issue to decide whether it meets the criteria for inclusion".

But it also seems like many issues are auto closed, still keeping the "evaluating" label. Will the team _actually_ evaluate the issue? Or am I realistically just waiting for this issue to also be auto closed?

I don't feel like implementing the feature unless there's a real chance of it being accepted.

**Comment by mdjermanovic** [comment]
@Skalman sorry for the delay! 

You can find more details on how we manage issues here:

https://eslint.org/docs/maintainer-guide/issues

In short, issues that have support from the team will be labeled as "accepted", meaning that we've agreed to implement the feature, and PRs are welcome. Unaccepted issues will be auto-closed after a certain period of inactivity, as it indicates that it isn't likely that the issue will gain support from the team.

What seems to be missing in this document (or rather needs to be updated) is our [RFC process](https://github.com/eslint/rfcs).

RuleTester is a part of ESLint core, so new features like the proposed one require going through the RFC process.

The first step is to open an issue to gather initial feedback, as you did.

I personally support this feature, because it provides an additional useful check for `invalid` tests, and also indirectly an additional set of `valid` tests (the `output` code) for free. However, I'd like to hear more opinions from team members before advising you to create an RFC pull request.

**Comment by anikethsaha** [comment]
Thanks @Skalman for clear explanation. 
I think this feature would be useful. 
Do we want all rules to be tested with this flag? Or if not, should there be any parameter(s) that would help us in finding which rule should be tested with this flag. 
And i suppose, there will be cases independent of each other for single rule where this flag would be used. 
I like the original proposed approach. IMO having this as default might be a breaking change. 
So i am thinking is, if we have `idempotent : true` (default `false`) , it would check for linting errors in the output. And it would simply fail the test if there is any otherwise pass the test. 

Wdyt.? I think it was similar to the original proposed solution.

**Comment by Skalman** [comment]
@mdjermanovic, @anikethsaha, thanks for the responses.

Enabling it by default would be a breaking change. Could be done earliest for 8.0.

There are probably two main alternatives:

   - `idempotent`
     - would only check that there are no further fixes available
     - could easily be on by default in 8.0
   - `outputIsValid`
     - would check that there are no more errors in the output
     - more useful and easier-to-understand name
     - I don't know how many tests this would break, so I don't know if it could be on by default in the future

**Comment by mdjermanovic** [comment]
@skalman you would prefer a name such as `outputIsValid` so that the (probably vast) majority of test cases can be set to `true` instead of `false`?

**Comment by Skalman** [comment]
@mdjermanovic, yes I prefer setting things to `true` rather than `false`, but perhaps you're misunderstanding what "idempotent" means? "Idempotent" means that running once and multiple times yield the same result.

So both `idempotent` and `outputIsValid` could be set to `true` for the majority of test cases.

It might be useful to try and figure out how vast that vast majority actually is. Is there a list of popular ESLint packages to test against? Going off of [npm popularity](https://www.npmjs.com/search?q=keywords%3Aeslintplugin&ranking=popularity), I'd suggest seeing how many existing test cases would fail out of these packages:

```
ESLint itself
eslint-plugin-react
eslint-plugin-import
eslint-plugin-jsx-a11y
eslint-plugin-prettier
eslint-plugin-react-hooks
@typescript-eslint/eslint-plugin
eslint-plugin-promise
eslint-plugin-node
eslint-plugin-jest
eslint-import-resolver-node
```

Is anything obvious missing from the above list?

**Comment by mdjermanovic** [comment]
I think it's safe to assume that the majority of test cases where the code has been fixed have `output` without linting errors, and I agree that `true` looks better than `false` as a common value.

My concern about the `outputIsValid` name is that it could be misinterpreted as "output doesn't have parsing errors", or even as "output is equivalent in behavior to the original code". Though, we're already using the term `valid` for the array of test cases where the rule isn't expected to report linting errors, so maybe this concern isn't valid :)

I'm not opposed to the `idempotent` check instead of the "lint-free" check, but it can be confusing as described in [this comment](https://github.com/eslint/eslint/issues/13631#issuecomment-683170219). The same code output after the second `--fix` doesn't have to mean that the second run didn't have linting errors. Also, why does it seem like a less-breaking change than the lint-free check?

**Comment by Skalman** [comment]
@mdjermanovic, both are breaking changes, but `idempotent` breaks fewer tests, though the difference isn't as big as I thought it would be.
I've experimented a bit, but only on ESLint's built-in rules. Anyway, here are some results (though I can't guarantee that my implementation is correct).

ESLint seems to contain 6170 "invalid" tests with an output.

## `outputIsValid`

Checks that the `output` is lint-free (that it passes as "valid").

**72 tests (1.17%) fail the output is valid check**
26 failing tests use a custom parser (under tests/fixtures/parsers/)

## `idempotent`

Checks that the `output` cannot be further fixed.

**58 tests (0.94%) fail the idempotent check**
24 failing tests use a custom parser (under tests/fixtures/parsers/)

----

Since the difference in breakage isn't that large, I'm leaning toward `outputIsValid`.

I'd assume that less well-maintained projects don't use custom parsers as much. It'd probably be good to experiment on popular plugins, but I don't think I'll be doing that any time soon.

**Comment by mdjermanovic** [comment]
Most, if not all of the fixture parsers will return the same "hard-coded" AST (taken from `babel-eslint` or `@typescript-eslint/parser` when the test was written) for both the `code` and the `output`, so those test cases are not relevant for this purpose. Plugins have a lot of tests where the real parser is used so the `output` code is actually parsed.

**Comment by mdjermanovic** [comment]
> Since the difference in breakage isn't that large, I'm leaning toward `outputIsValid`.

I'm also in favor of `outputIsValid`. but we could also describe `idempotent` in the [Alternatives](https://github.com/eslint/rfcs/blob/master/templates/design.md#alternatives) section and hear more opinions from the community.

@Skalman would you like to create an [RFC](https://github.com/eslint/rfcs) for this change?
